students/EricAdams/session04/trigrams.py

```diff
@@ -1,101 +1,166 @@
+ At this point this script will write the final output
+ The final output file will be a mutated form of the input text file.
+ """
 import random
+# import string
+
+# raw_input_list = []
+input_string = []
+# initial_trigram_key = tuple()
+# output_list = []
+
+# Make these strings instead of lists for the str.translate method
+
+# List of chars to remove or replace with a space from input text
+# unwanted_punc = ['-',
+#                  "'",
+#                  ',',
+#                  '.',
+#                  ')',
+#                  '(',
+#                  '"',
+#                  '?',
+#                  ';',
+#                  '\n']
+
+# substitution_chars = [' ',
+#                       ' ',
+#                       ' ',
+#                       ' ',
+#                       ' ',
+#                       ' ',
+#                       ' ',
+#                       ' ',
+#                       ' ',
+#                       ' ']
+substitution_chars = ' ' * 10
+unwanted_punc = '-' + "'" + ',' + '.' + ')' + '(' + '"' + '?' + ';' + '\n'
+
+
+def replace_unwanted_punctuation(raw_input_string, unwanted_punc,
+                                 substitution_chars):
+    """ Build up an input string with unwanted punctuation replaced
+    in accordance with a string of chars
+    :param1 = a string with punctuation
+    :param2 = a string of unwanted punctuation
+    :param3 = a string of char to substitue for the unwanted punctuation
+    :return a string free of unwanted punctuation
+    """
+    # Now that the read in file is one string. Use str.translate
+    # for i, x in enumerate(raw_input_list):
+    #     if x in replace_punc:
+    #         input_list.append(' ')
+    #     else:
+    #         input_list.append(x)
+
+    # maketrans returns a table
+    transition_table = raw_input_string.maketrans(
+        unwanted_punc, substitution_chars)
+    input_string = raw_input_string.translate(transition_table)
+    return input_string
+
+
+def create_the_trigram(input_string):
+    """ Create a trigram from a list.
+    :param1 = a list
+    :return = a dictionary, (trigram)
+    """
+    trigram = {}
+    # we need a pair hence the len(input_list) - 2
+    #   instead of len(input_list) - 1
+    for i in range(len(input_string) - 2):
+        # must be a tuple in order to be a key in trigram
+        trigram_key = tuple(input_string[i:i + 2])
+        trigram_followers = input_string[i + 2]
+        # this is so nifty, I had to use it
+        trigram.setdefault(trigram_key, []).append(trigram_followers)
+    return trigram
+
+
+def initial_output_list(trigram):
+    """Initialize the output list with three elements derived from a
+    random trigram key
+    :param = dictionary, (trigram)
+    :return list
+    """
+    # Remove - 1
+    # random_number = random.randrange(0, len(trigram) - 1)
+    random_number = random.randrange(0, len(trigram))
+    for i, trigram_key in enumerate(trigram):
```

Comment:
kinda klunky, but it works :-)

As we talked about in class, I like:

initial_trigram_key = random.choice(list(trigram))

but that does require making a list of all the keys ahead of time -- could be memory intensive.